engine/execution/ingestion/engine.go

```diff
@@ -1113,6 +1113,26 @@ func (e *Engine) saveExecutionResults(
 		startState = endState
 	}
 
+	// before persisting state commitment, check whether this block has been sealed already.
+	// if the block has been sealed, it means some other execution node has executed this block
+	// already, and its result has already been accepted by the consensus nodes. In that
+	// case, we need to double check whether our execution result is consistent with the sealed
+	// result. If inconsistent, it means there is a fork in the execution results, which
+	// would be abnormal. we'd better stop the node in this case.
+	commit, err := e.state.AtBlockID(blockID).Commit()
+	if err == nil {
+		// the block has been sealed
+		isExecutionEndStateInconsistentWithSealedState := !bytes.Equal(commit, endState)
+		if isExecutionEndStateInconsistentWithSealedState {
```

Comment:
This might not be as useful as we thought, because in most cases the execution node executes the latest block, which is not sealed or even finalized, so the check always skipped. So this would happen if two independent events happen at the same time (which is very unlikely):
- The node is on an execution result fork that is different from the sealed execution fork)
- The node is failing behind

Instead, we should just check if the latest sealed block's seal contains the same state commitment as the one I produced. Since last sealed block only get updated only when blocks are finalized, we could listen to the block finalized event to perform the check.

Reply:
> Instead, we should just check if the latest sealed block's seal contains the same state commitment as the one I produced. 

great suggestion. That would be a much more powerful check increasing the confidence in our execution nodes 🤩